File: health_sync/providers/eightsleep.py
from __future__ import annotations

from datetime import UTC, datetime, timedelta

import requests

from ..config import LoadedConfig, require_str
from ..db import HealthSyncDb
from .runtime import stable_json, sync_resource, token_expiring_soon, token_extra
from ..util import iso_to_dt, parse_yyyy_mm_dd, request_json, sha256_hex, utc_now_iso


EIGHTSLEEP_PROVIDER = "eightsleep"


def _eightsleep_auth_headers() -> dict[str, str]:
    return {
        "Content-Type": "application/json",
        "User-Agent": "health-sync/0.1 (+local sqlite cache)",
        "Accept": "application/json",
    }


def _eightsleep_api_headers(access_token: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json",
        "User-Agent": "health-sync/0.1 (+local sqlite cache)",
        "Accept": "application/json",
    }


def _parse_date(s: str | None) -> datetime | None:
    if not s:
        return None
    raw = s.strip()
    if not raw:
        return None
    try:
        return iso_to_dt(raw)
    except Exception:  # noqa: BLE001
        pass
    if len(raw) >= 10 and raw[4:5] == "-" and raw[7:8] == "-":
        try:
            return parse_yyyy_mm_dd(raw[:10])
        except Exception:  # noqa: BLE001
            return None
    return None


def _eightsleep_refresh_if_needed(db: HealthSyncDb, cfg: LoadedConfig, sess: requests.Session) -> str:
    # Static token in config takes precedence.
    static_access_token = cfg.config.eightsleep.access_token
    if static_access_token:
        return static_access_token

    tok = db.get_oauth_token(EIGHTSLEEP_PROVIDER)
    if tok:
        if not token_expiring_soon(tok.get("expires_at"), skew_seconds=60):
            return tok["access_token"]

    auth_url = cfg.config.eightsleep.auth_url.rstrip("/")
    email = require_str(cfg, cfg.config.eightsleep.email, key="eightsleep.email")
    password = require_str(cfg, cfg.config.eightsleep.password, key="eightsleep.password")
    client_id = require_str(cfg, cfg.config.eightsleep.client_id, key="eightsleep.client_id")
    client_secret = require_str(cfg, cfg.config.eightsleep.client_secret, key="eightsleep.client_secret")

    token = request_json(
        sess,
        "POST",
        auth_url,
        headers=_eightsleep_auth_headers(),
        json_data={
            "client_id": client_id,
            "client_secret": client_secret,
            "grant_type": "password",
            "username": email,
            "password": password,
        },
    )

    access_token = token["access_token"]
    expires_at = None
    expires_in = token.get("expires_in")
    try:
        if expires_in is not None:
            expires_at = (datetime.now(UTC) + timedelta(seconds=int(float(expires_in)))).replace(microsecond=0).isoformat().replace("+00:00", "Z")
    except Exception:  # noqa: BLE001
        expires_at = None

    db.set_oauth_token(
        provider=EIGHTSLEEP_PROVIDER,
        access_token=access_token,
        refresh_token=None,
        token_type="Bearer",
        scope=None,
        expires_at=expires_at,
        extra=token_extra(token, excluded=("access_token", "expires_in")),
    )
    return access_token


def _trend_start_date(db: HealthSyncDb, cfg: LoadedConfig) -> str:
    state = db.get_sync_state(provider=EIGHTSLEEP_PROVIDER, resource="trends")
    overlap_days = int(cfg.config.eightsleep.overlap_days)

    if state and state.watermark:
        dt = _parse_date(state.watermark)
        if dt is not None:
            return (dt.date() - timedelta(days=overlap_days)).isoformat()
    return cfg.config.eightsleep.start_date


def eightsleep_sync(db: HealthSyncDb, cfg: LoadedConfig) -> None:
    sess = requests.Session()
    access_token = _eightsleep_refresh_if_needed(db, cfg, sess)
    headers = _eightsleep_api_headers(access_token)

    client_api_url = cfg.config.eightsleep.client_api_url.rstrip("/")
    tz = cfg.config.eightsleep.timezone or "UTC"
    today = datetime.now(UTC).date().isoformat()

    print("Syncing Eight Sleep...")

    me_resp = request_json(sess, "GET", f"{client_api_url}/users/me", headers=headers)
    me_user = me_resp.get("user") if isinstance(me_resp, dict) else None
    me_id = str((me_user or {}).get("id") or "me")

    with sync_resource(db, provider=EIGHTSLEEP_PROVIDER, resource="users_me") as run_users_me:
        op = db.upsert_record(
            provider=EIGHTSLEEP_PROVIDER,
            resource="users_me",
            record_id=me_id,
            payload=me_resp,
            start_time=None,
            end_time=None,
            source_updated_at=utc_now_iso(),
        )
        run_users_me.add_upsert(op)
        db.set_sync_state(provider=EIGHTSLEEP_PROVIDER, resource="users_me", watermark=utc_now_iso())

    user_ids: set[str] = set()
    if me_id != "me":
        user_ids.add(me_id)

    device_ids = (me_user or {}).get("devices")
    device_id = str(device_ids[0]) if isinstance(device_ids, list) and device_ids else None
    if device_id:
        device_resp = request_json(sess, "GET", f"{client_api_url}/devices/{device_id}", headers=headers)
        result = device_resp.get("result") if isinstance(device_resp, dict) else {}
        if isinstance(result, dict):
            left_id = result.get("leftUserId")
            right_id = result.get("rightUserId")
            if left_id:
                user_ids.add(str(left_id))
            if right_id:
                user_ids.add(str(right_id))
            away_sides = result.get("awaySides")
            if isinstance(away_sides, dict):
                for uid in away_sides.values():
                    if uid:
                        user_ids.add(str(uid))

        with sync_resource(db, provider=EIGHTSLEEP_PROVIDER, resource="devices") as run_devices:
            op = db.upsert_record(
                provider=EIGHTSLEEP_PROVIDER,
                resource="devices",
                record_id=device_id,
                payload=device_resp,
                start_time=None,
                end_time=None,
                source_updated_at=utc_now_iso(),
            )
            run_devices.add_upsert(op)
            db.set_sync_state(provider=EIGHTSLEEP_PROVIDER, resource="devices", watermark=utc_now_iso())

    total_profiles = 0
    total_trends = 0
    from_date = _trend_start_date(db, cfg)

    with db.sync_run(provider=EIGHTSLEEP_PROVIDER, resource="users") as run_users:
        with db.sync_run(provider=EIGHTSLEEP_PROVIDER, resource="trends") as run_trends:
            with db.transaction():
                for user_id in sorted(user_ids):
                    profile_resp = request_json(sess, "GET", f"{client_api_url}/users/{user_id}", headers=headers)
                    op = db.upsert_record(
                        provider=EIGHTSLEEP_PROVIDER,
                        resource="users",
                        record_id=str(user_id),
                        payload=profile_resp,
                        start_time=None,
                        end_time=None,
                        source_updated_at=utc_now_iso(),
                    )
                    run_users.add_upsert(op)
                    total_profiles += 1

                    trend_resp = request_json(
                        sess,
                        "GET",
                        f"{client_api_url}/users/{user_id}/trends",
                        headers=headers,
                        params={
                            "tz": tz,
                            "from": from_date,
                            "to": today,
                            "include-main": "false",
                            "include-all-sessions": "true",
                            "model-version": "v2",
                        },
                    )

                    days = trend_resp.get("days") if isinstance(trend_resp, dict) else None
                    if not isinstance(days, list):
                        continue
                    for day in days:
                        if not isinstance(day, dict):
                            continue
                        day_key = day.get("day")
                        rid = f"{user_id}:{day_key}" if day_key else f"{user_id}:{sha256_hex(stable_json(day))}"
                        start_time = day.get("day") or day.get("presenceStart")
                        end_time = day.get("presenceEnd")
                        source_updated_at = day.get("updatedAt") or day.get("presenceStart") or start_time
                        op_trend = db.upsert_record(
                            provider=EIGHTSLEEP_PROVIDER,
                            resource="trends",
                            record_id=rid,
                            payload=day,
                            start_time=start_time,
                            end_time=end_time,
                            source_updated_at=source_updated_at,
                        )
                        run_trends.add_upsert(op_trend)
                        total_trends += 1

                db.set_sync_state(provider=EIGHTSLEEP_PROVIDER, resource="users", watermark=utc_now_iso())
                db.set_sync_state(provider=EIGHTSLEEP_PROVIDER, resource="trends", watermark=utc_now_iso())

    print(f"Eight Sleep sync complete ({total_profiles} user profiles, {total_trends} trend rows).")

File: src/providers/eightsleep.js
import {
  parseYYYYMMDD,
  requestJson,
  sha256Hex,
  utcNowIso,
} from '../util.js';

function dateToYYYYMMDD(date) {
  const y = date.getUTCFullYear();
  const m = String(date.getUTCMonth() + 1).padStart(2, '0');
  const d = String(date.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

function isoToDate(value) {
  if (!value) {
    return null;
  }
  const parsed = new Date(value);
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

function tokenExpiredSoon(expiresAtIso, skewSeconds = 60) {
  if (!expiresAtIso) {
    return true;
  }
  const expiryMs = Date.parse(expiresAtIso);
  if (Number.isNaN(expiryMs)) {
    return true;
  }
  return expiryMs <= (Date.now() + skewSeconds * 1000);
}

function tokenExtra(payload) {
  const out = {};
  for (const [key, value] of Object.entries(payload || {})) {
    if (key === 'access_token' || key === 'expires_in') {
      continue;
    }
    out[key] = value;
  }
  return out;
}

async function eightsleepRefreshIfNeeded(db, cfg) {
  if (cfg.access_token) {
    db.setOAuthToken('eightsleep', {
      accessToken: String(cfg.access_token),
      refreshToken: null,
      tokenType: 'Bearer',
      scope: null,
      expiresAt: null,
      extra: { method: 'static_access_token' },
    });
    return String(cfg.access_token);
  }

  const existing = db.getOAuthToken('eightsleep');
  if (existing && !tokenExpiredSoon(existing.expiresAt)) {
    return existing.accessToken;
  }

  if (!cfg.email || !cfg.password) {
    throw new Error('Missing [eightsleep].email or [eightsleep].password, or provide [eightsleep].access_token.');
  }

  const tokenPayload = await requestJson(cfg.auth_url || 'https://auth-api.8slp.net/v1/tokens', {
    method: 'POST',
    json: {
      client_id: cfg.client_id,
      client_secret: cfg.client_secret,
      grant_type: 'password',
      username: cfg.email,
      password: cfg.password,
    },
  });

  if (!tokenPayload?.access_token) {
    throw new Error(`Eight Sleep auth failed: ${JSON.stringify(tokenPayload)}`);
  }

  const expiresIn = Number.parseInt(String(tokenPayload.expires_in ?? 0), 10) || 0;
  const expiresAt = expiresIn > 0 ? new Date(Date.now() + expiresIn * 1000).toISOString().replace(/\.\d{3}Z$/, 'Z') : null;

  db.setOAuthToken('eightsleep', {
    accessToken: String(tokenPayload.access_token),
    refreshToken: null,
    tokenType: tokenPayload.token_type || 'Bearer',
    scope: tokenPayload.scope || null,
    expiresAt,
    extra: tokenExtra(tokenPayload),
  });

  return String(tokenPayload.access_token);
}

function trendsFromDate(db, cfg) {
  const overlapDays = Math.max(0, Number.parseInt(String(cfg.overlap_days ?? 2), 10) || 0);
  const state = db.getSyncState('eightsleep', 'trends');
  if (!state?.watermark) {
    return String(cfg.start_date || '2010-01-01');
  }

  const wmDate = isoToDate(state.watermark);
  if (!wmDate) {
    return String(cfg.start_date || '2010-01-01');
  }

  const fromDate = new Date(Date.UTC(
    wmDate.getUTCFullYear(),
    wmDate.getUTCMonth(),
    wmDate.getUTCDate(),
    0,
    0,
    0,
  ));
  fromDate.setUTCDate(fromDate.getUTCDate() - overlapDays);
  return dateToYYYYMMDD(fromDate);
}

function collectDeviceUserIds(payload) {
  const ids = [];
  const root = payload?.result && typeof payload.result === 'object' ? payload.result : payload;
  if (!root || typeof root !== 'object') {
    return ids;
  }

  for (const key of ['leftUserId', 'rightUserId']) {
    if (root[key] !== undefined && root[key] !== null) {
      ids.push(String(root[key]));
    }
  }

  const awaySides = root.awaySides;
  if (awaySides && typeof awaySides === 'object') {
    for (const value of Object.values(awaySides)) {
      if (value !== undefined && value !== null && String(value).trim()) {
        ids.push(String(value));
      }
    }
  }

  return ids;
}

function extractTrendEntries(payload) {
  if (Array.isArray(payload?.result)) {
    return payload.result;
  }
  if (Array.isArray(payload?.trends)) {
    return payload.trends;
  }
  if (Array.isArray(payload?.days)) {
    return payload.days;
  }
  return [];
}

function trendRecordId(userId, trend) {
  if (trend?.day) {
    return `${userId}:${String(trend.day)}`;
  }
  return `${userId}:${sha256Hex(JSON.stringify(trend))}`;
}

async function eightsleepAuth(db, config, helpers) {
  const cfg = helpers.configFor('eightsleep');
  await eightsleepRefreshIfNeeded(db, cfg);
  console.log('Eight Sleep authorization succeeded.');
}

async function eightsleepSync(db, config, helpers) {
  const cfg = helpers.configFor('eightsleep');
  const accessToken = await eightsleepRefreshIfNeeded(db, cfg);

  const apiBase = String(cfg.client_api_url || 'https://client-api.8slp.net/v1').replace(/\/$/, '');
  const headers = {
    Authorization: `Bearer ${accessToken}`,
  };

  let mePayload;
  let meUserId = 'me';
  let meUser = null;

  await db.syncRun('eightsleep', 'users_me', async () => {
    await db.transaction(async () => {
      mePayload = await requestJson(`${apiBase}/users/me`, { headers });
      meUser = mePayload?.user && typeof mePayload.user === 'object' ? mePayload.user : null;
      meUserId = String(meUser?.id || mePayload?.id || 'me');
      const nowIso = utcNowIso();
      db.upsertRecord({
        provider: 'eightsleep',
        resource: 'users_me',
        recordId: meUserId,
        startTime: null,
        endTime: null,
        sourceUpdatedAt: nowIso,
        payload: mePayload,
      });
      db.setSyncState('eightsleep', 'users_me', {
        watermark: nowIso,
      });
    });
  });

  const userIds = new Set();
  if (meUserId !== 'me') {
    userIds.add(meUserId);
  }

  await db.syncRun('eightsleep', 'devices', async () => {
    await db.transaction(async () => {
      const devices = Array.isArray(meUser?.devices)
        ? meUser.devices
        : Array.isArray(mePayload?.devices)
          ? mePayload.devices
          : [];
      const nowIso = utcNowIso();

      for (const deviceEntry of devices) {
        const deviceId = typeof deviceEntry === 'object' && deviceEntry !== null
          ? deviceEntry.id || deviceEntry.deviceId
          : deviceEntry;
        if (!deviceId) {
          continue;
        }

        const payload = await requestJson(`${apiBase}/devices/${encodeURIComponent(String(deviceId))}`, {
          headers,
        });

        db.upsertRecord({
          provider: 'eightsleep',
          resource: 'devices',
          recordId: String(deviceId),
          startTime: null,
          endTime: null,
          sourceUpdatedAt: nowIso,
          payload,
        });

        for (const uid of collectDeviceUserIds(payload)) {
          userIds.add(uid);
        }
      }

      db.setSyncState('eightsleep', 'devices', {
        watermark: nowIso,
      });
    });
  });

  await db.syncRun('eightsleep', 'users', async () => {
    await db.transaction(async () => {
      const nowIso = utcNowIso();
      for (const userId of userIds) {
        const payload = await requestJson(`${apiBase}/users/${encodeURIComponent(String(userId))}`, {
          headers,
        });
        db.upsertRecord({
          provider: 'eightsleep',
          resource: 'users',
          recordId: String(userId),
          startTime: null,
          endTime: null,
          sourceUpdatedAt: nowIso,
          payload,
        });
      }

      db.setSyncState('eightsleep', 'users', {
        watermark: nowIso,
      });
    });
  });

  await db.syncRun('eightsleep', 'trends', async () => {
    await db.transaction(async () => {
      const timezone = String(cfg.timezone || 'UTC');
      const fromDate = trendsFromDate(db, cfg);
      const todayDate = dateToYYYYMMDD(new Date());
      const nowIso = utcNowIso();

      for (const userId of userIds) {
        const payload = await requestJson(`${apiBase}/users/${encodeURIComponent(String(userId))}/trends`, {
          headers,
          params: {
            tz: timezone,
            from: fromDate,
            to: todayDate,
            'include-main': 'false',
            'include-all-sessions': 'true',
            'model-version': 'v2',
          },
        });

        const entries = extractTrendEntries(payload);
        for (const trend of entries) {
          const startTime = trend?.day || trend?.presenceStart || null;
          const sourceUpdatedAt = trend?.updatedAt || trend?.presenceStart || startTime;
          db.upsertRecord({
            provider: 'eightsleep',
            resource: 'trends',
            recordId: trendRecordId(String(userId), trend),
            startTime,
            endTime: trend?.presenceEnd || null,
            sourceUpdatedAt,
            payload: trend,
          });
        }
      }

      db.setSyncState('eightsleep', 'trends', {
        watermark: nowIso,
      });
    });
  });
}

const eightsleepProvider = {
  id: 'eightsleep',
  source: 'builtin',
  description: 'Eight Sleep user/device and trend data sync',
  supportsAuth: true,
  auth: eightsleepAuth,
  sync: eightsleepSync,
};

export default eightsleepProvider;
